# Lab1-Assignment

Copyright: Vrije Universiteit Amsterdam, Faculty of Humanities, CLTL

This notebook describes the assignment for Lab 1 of the text mining course. 

**Points**: each exercise is prefixed with the number of points you can obtain for the exercise.

We assume you have worked through the following notebooks:
* **Lab1.1-introduction**
* **Lab1.2-introduction-to-NLTK**
* **Lab1.3-introduction-to-spaCy** 

In this assignment, you will process an English text (**Lab1-apple-samsung-example.txt**) with both NLTK and spaCy and discuss the similarities and differences.

## Credits
The notebooks in this block have been originally created by [Marten Postma](https://martenpostma.github.io). Adaptations were made by [Filip Ilievski](http://ilievski.nl).

## Tip: how to read a file from disk
Let's open the file **Lab1-apple-samsung-example.txt** from disk.

In [1]:
from pathlib import Path

In [2]:
cur_dir = Path().resolve() # this should provide you with the folder in which this notebook is placed
path_to_file = Path.joinpath(cur_dir, 'Lab1-apple-samsung-example.txt')
print(path_to_file)
print('does path exist? ->', Path.exists(path_to_file))

/Users/demytran/Desktop/textmining/ba-text-mining/lab_sessions/lab1/Lab1-apple-samsung-example.txt
does path exist? -> True


If the output from the code cell above states that **does path exist? -> False**, please check that the file **Lab1-apple-samsung-example.txt** is in the same directory as this notebook.

In [3]:
with open(path_to_file) as infile:
    text = infile.read()

print('number of characters', len(text))

number of characters 1139


## [total points: 4] Exercise 1: NLTK
In this exercise, we use NLTK to apply **Part-of-speech (POS) tagging**, **Named Entity Recognition (NER)**, and **Constituency parsing**. The following code snippet already performs sentence splitting and tokenization. 

In [4]:
import nltk
from nltk.tokenize import sent_tokenize
from nltk import word_tokenize

In [5]:
sentences_nltk = sent_tokenize(text)

In [6]:
tokens_per_sentence = []
for sentence_nltk in sentences_nltk:
    sent_tokens = word_tokenize(sentence_nltk)
    tokens_per_sentence.append(sent_tokens)

We will use lists to keep track of the output of the NLP tasks. We can hence inspect the output for each task using the index of the sentence.

In [9]:
sent_id = 1
print('SENTENCE', sentences_nltk[sent_id])
print('TOKENS', tokens_per_sentence[sent_id])

SENTENCE The six phones and tablets affected are the Galaxy S III, running the new Jelly Bean system, the Galaxy Tab 8.9 Wifi tablet, the Galaxy Tab 2 10.1, Galaxy Rugby Pro and Galaxy S III mini.
TOKENS ['The', 'six', 'phones', 'and', 'tablets', 'affected', 'are', 'the', 'Galaxy', 'S', 'III', ',', 'running', 'the', 'new', 'Jelly', 'Bean', 'system', ',', 'the', 'Galaxy', 'Tab', '8.9', 'Wifi', 'tablet', ',', 'the', 'Galaxy', 'Tab', '2', '10.1', ',', 'Galaxy', 'Rugby', 'Pro', 'and', 'Galaxy', 'S', 'III', 'mini', '.']


### [point: 1] Exercise 1a: Part-of-speech (POS) tagging
Use `nltk.pos_tag` to perform part-of-speech tagging on each sentence.

Use `print` to **show** the output in the notebook (and hence also in the exported PDF!).

In [ ]:
pos_tags_per_sentence = []
for tokens in tokens_per_sentence:
    pos_tags = nltk.pos_tag(tokens)
    pos_tags_per_sentence.append(pos_tags)
    print(pos_tags)
    print()


In [15]:
print(pos_tags_per_sentence)

[[('https', 'NN'), (':', ':'), ('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ'), ('Documents', 'NNS'), ('filed', 'VBN'), ('to', 'TO'), ('the', 'DT'), ('San', 'NNP'), ('Jose', 'NNP'), ('federal', 'JJ'), ('court', 'NN'), ('in', 'IN'), ('California', 'NNP'), ('on', 'IN'), ('November', 'NNP'), ('23', 'CD'), ('list', 'NN'), ('six', 'CD'), ('Samsung', 'NNP'), ('products', 'NNS'), ('running', 'VBG'), ('the', 'DT'), ('``', '``'), ('Jelly', 'RB'), ('Bean', 'NNP'), ("''", "''"), ('and', 'CC'), ('``', '``'), ('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP'), ("''", "''"), ('operating', 'VBG'), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), ('Apple', 'NNP'), ('claims', 'VBZ'), ('infringe', 'VB'), ('its', 'PRP$'), ('patents', 'NNS'), ('.', '.')], [('The', 'DT'), ('six', 'CD'), ('phones', 'NNS'), ('and', 'CC'), ('tablets', 'NNS'), ('affected', 'VBN'), ('are', 'VBP'), ('the', 'DT'), ('Galaxy', 'NNP'), ('S', 'NNP'), ('III', 'NN

### [point: 1] Exercise 1b: Named Entity Recognition (NER)
Use `nltk.chunk.ne_chunk` to perform Named Entity Recognition (NER) on each sentence.

Use `print` to **show** the output in the notebook (and hence also in the exported PDF!).

In [ ]:
from nltk.chunk import ne_chunk

ner_tags_per_sentence = []
for pos_tags in pos_tags_per_sentence:
    ner_tree = ne_chunk(pos_tags)
    ner_tags_per_sentence.append(ner_tree)
    print(ner_tree)
    print()


In [19]:
print(ner_tags_per_sentence)

[Tree('S', [('https', 'NN'), (':', ':'), ('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ'), ('Documents', 'NNS'), ('filed', 'VBN'), ('to', 'TO'), ('the', 'DT'), Tree('ORGANIZATION', [('San', 'NNP'), ('Jose', 'NNP')]), ('federal', 'JJ'), ('court', 'NN'), ('in', 'IN'), Tree('GPE', [('California', 'NNP')]), ('on', 'IN'), ('November', 'NNP'), ('23', 'CD'), ('list', 'NN'), ('six', 'CD'), Tree('ORGANIZATION', [('Samsung', 'NNP')]), ('products', 'NNS'), ('running', 'VBG'), ('the', 'DT'), ('``', '``'), ('Jelly', 'RB'), Tree('GPE', [('Bean', 'NNP')]), ("''", "''"), ('and', 'CC'), ('``', '``'), ('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP'), ("''", "''"), ('operating', 'VBG'), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), Tree('PERSON', [('Apple', 'NNP')]), ('claims', 'VBZ'), ('infringe', 'VB'), ('its', 'PRP$'), ('patents', 'NNS'), ('.', '.')]), Tree('S', [('The', 'DT'), ('six', 'CD'), ('phones', 'NNS'), ('and', 'CC'),

### [points: 2] Exercise 1c: Constituency parsing
Use the `nltk.RegexpParser` to perform constituency parsing on each sentence.

Use `print` to **show** the output in the notebook (and hence also in the exported PDF!).

In [21]:
constituent_parser = nltk.RegexpParser('''
NP: {<DT>? <JJ>* <NN>*} # NP
P: {<IN>}           # Preposition
V: {<V.*>}          # Verb
PP: {<P> <NP>}      # PP -> P NP
VP: {<V> <NP|PP>*}  # VP -> V (NP|PP)*''')

In [ ]:
constituency_output_per_sentence = []
for pos_tags in pos_tags_per_sentence:
    parse_tree = constituent_parser.parse(pos_tags)
    constituency_output_per_sentence.append(parse_tree)
    print(parse_tree)
    print()


In [23]:
print(constituency_output_per_sentence)

[Tree('S', [Tree('NP', [('https', 'NN')]), (':', ':'), Tree('NP', [('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ')]), ('Documents', 'NNS'), Tree('VP', [Tree('V', [('filed', 'VBN')])]), ('to', 'TO'), Tree('NP', [('the', 'DT')]), ('San', 'NNP'), ('Jose', 'NNP'), Tree('NP', [('federal', 'JJ'), ('court', 'NN')]), Tree('P', [('in', 'IN')]), ('California', 'NNP'), Tree('P', [('on', 'IN')]), ('November', 'NNP'), ('23', 'CD'), Tree('NP', [('list', 'NN')]), ('six', 'CD'), ('Samsung', 'NNP'), ('products', 'NNS'), Tree('VP', [Tree('V', [('running', 'VBG')]), Tree('NP', [('the', 'DT')])]), ('``', '``'), ('Jelly', 'RB'), ('Bean', 'NNP'), ("''", "''"), ('and', 'CC'), ('``', '``'), ('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP'), ("''", "''"), Tree('VP', [Tree('V', [('operating', 'VBG')])]), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), ('Apple', 'NNP'), Tree('VP', [Tree('V', [('claims', 'VBZ')])]), Tree('VP', [Tree('V', [

Augment the RegexpParser so that it also detects Named Entity Phrases (NEP), e.g., that it detects *Galaxy S III* and *Ice Cream Sandwich*

In [ ]:
constituent_parser_v2 = nltk.RegexpParser('''
NEP: {<NNP>+(<CD>|<NNP>|<NN>)*}   # Named Entity Phrase: proper nouns + optional nums/nouns
NP: {<DT>? <JJ>* <NN>*}            # NP
P: {<IN>}                           # Preposition
V: {<V.*>}                          # Verb
PP: {<P> <NP>}                      # PP -> P NP
VP: {<V> <NP|PP>*}                  # VP -> V (NP|PP)*''')


In [ ]:
constituency_v2_output_per_sentence = []
for pos_tags in pos_tags_per_sentence:
    parse_tree_v2 = constituent_parser_v2.parse(pos_tags)
    constituency_v2_output_per_sentence.append(parse_tree_v2)
    print(parse_tree_v2)
    print()


In [27]:
print(constituency_v2_output_per_sentence)

[Tree('S', [Tree('NP', [('https', 'NN')]), (':', ':'), Tree('NP', [('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ')]), ('Documents', 'NNS'), Tree('VP', [Tree('V', [('filed', 'VBN')])]), ('to', 'TO'), Tree('NP', [('the', 'DT')]), Tree('NEP', [('San', 'NNP'), ('Jose', 'NNP')]), Tree('NP', [('federal', 'JJ'), ('court', 'NN')]), Tree('P', [('in', 'IN')]), Tree('NEP', [('California', 'NNP')]), Tree('P', [('on', 'IN')]), Tree('NEP', [('November', 'NNP')]), ('23', 'CD'), Tree('NP', [('list', 'NN')]), ('six', 'CD'), Tree('NEP', [('Samsung', 'NNP')]), ('products', 'NNS'), Tree('VP', [Tree('V', [('running', 'VBG')]), Tree('NP', [('the', 'DT')])]), ('``', '``'), ('Jelly', 'RB'), Tree('NEP', [('Bean', 'NNP')]), ("''", "''"), ('and', 'CC'), ('``', '``'), Tree('NEP', [('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP')]), ("''", "''"), Tree('VP', [Tree('V', [('operating', 'VBG')])]), ('systems', 'NNS'), (',', ','), ('which', 'WDT'

## [total points: 1] Exercise 2: spaCy
Use Spacy to process the same text as you analyzed with NLTK.

In [28]:
import spacy
nlp = spacy.load('en_core_web_sm')

In [ ]:
doc = nlp(text)
sents = list(doc.sents)

for i, sent in enumerate(sents):
    print(f"--- Sentence {i} ---")
    print(sent.text)
    print()
    for token in sent:
        print(f"  {token.text:<20} pos: {token.pos_:<8} tag: {token.tag_:<6} dep: {token.dep_}")
    ents_in_sent = [e for e in doc.ents if e.start >= sent.start and e.end <= sent.end]
    if ents_in_sent:
        print("  Named Entities:")
        for ent in ents_in_sent:
            print(f"    {ent.text:<30} {ent.label_}")
    print()


small tip: You can use **sents = list(doc.sents)** to be able to use the index to access a sentence like **sents[2]** for the third sentence.


In [ ]:
# Sentence splitting comparison
print(f"NLTK identified {len(sentences_nltk)} sentences")
print(f"spaCy identified {len(sents)} sentences")
print()
for i, (ns, ss) in enumerate(zip(sentences_nltk, sents)):
    match = "==" if ns.strip() == str(ss).strip() else "!="
    print(f"Sentence {i} {match}")
    print(f"  NLTK : {ns[:90]}")
    print(f"  spaCy: {str(ss)[:90]}")
    print()


In [ ]:
# POS tag comparison for sentence 1 (NLTK vs spaCy)
nltk_tags  = pos_tags_per_sentence[1]
spacy_tags = [(token.text, token.tag_) for token in sents[1]]

print(f"{'Token':<20} {'NLTK':<10} {'spaCy':<10} {'Match?'}")
print("-" * 55)
for j in range(max(len(nltk_tags), len(spacy_tags))):
    n_word, n_tag = nltk_tags[j]  if j < len(nltk_tags)  else ("-", "-")
    s_word, s_tag = spacy_tags[j] if j < len(spacy_tags) else ("-", "-")
    flag = "  " if n_tag == s_tag else "!="
    print(f"{s_word:<20} {n_tag:<10} {s_tag:<10} {flag}")


In [ ]:
# NER comparison for sentence 3
sent_id = 3
print("=== NLTK NER (sentence 3) ===")
print(ner_tags_per_sentence[sent_id])
print()
print("=== spaCy NER (sentence 3) ===")
for ent in sents[sent_id].ents:
    print(f"  {ent.text:<30} {ent.label_}")


In [ ]:
# Constituency (NLTK) vs Dependency (spaCy) for sentence 3
sent_id = 3

print("=== NLTK Constituency Parse ===")
print(constituency_output_per_sentence[sent_id])
print()

print("=== spaCy Dependency Parse ===")
for token in sents[sent_id]:
    print(f"  {token.text:<15} --[{token.dep_:<10}]--> {token.head.text}")


## [total points: 7] Exercise 3: Comparison NLTK and spaCy
We will now compare the output of NLTK and spaCy, i.e., in what do they differ?

### [points: 3] Exercise 3a: Part of speech tagging
Compare the output from NLTK and spaCy regarding part of speech tagging.

* To compare, you probably would like to compare sentence per sentence. Describe if the sentence splitting is different for NLTK than for spaCy. If not, where do they differ?
* After checking the sentence splitting, select a sentence for which you expect interesting results and perhaps differences. Motivate your choice.
* Compare the output in `token.tag` from spaCy to the part of speech tagging from NLTK for each token in your selected sentence. Are there any differences? This is not a trick question; it is possible that there are no differences.

### Answer

**Sentence splitting:**
Both NLTK and spaCy split the text into 6 sentences and agree on all boundaries.
NLTK uses the Punkt tokenizer (rule-based), while spaCy derives sentence boundaries from
its dependency parser, but for this text the results are identical.

**Selected sentence (index 1):**
*"The six phones and tablets affected are the Galaxy S III, running the new Jelly Bean
system, the Galaxy Tab 8.9 Wifi tablet, the Galaxy Tab 2 10.1, Galaxy Rugby Pro and
Galaxy S III mini."*

This sentence is a good test case because it contains product names with mixed
proper-noun/number tokens (Galaxy S III, Tab 8.9, Tab 2 10.1), which are tricky to tag.

**POS tag comparison:**
Both NLTK and spaCy use the Penn Treebank tagset, so tags are directly comparable.
The outputs are largely identical:
- Product name tokens (*Galaxy*, *Rugby*, *Wifi*, *Tab*) → NNP in both.
- Numbers (*8.9*, *10.1*, *2*, *six*) → CD in both.
- *III* → NNP (treated as a proper noun) in both.
- *mini* is the main potential difference: NLTK may tag it as JJ (adjective) while
  spaCy may tag it as NN or NNP depending on context.

Overall the two taggers agree on this sentence, with only minor edge-case differences.


### [points: 2] Exercise 3b: Named Entity Recognition (NER)
* Describe differences between the output from NLTK and spaCy for Named Entity Recognition. Which one do you think performs better?

### Answer

**Differences between NLTK and spaCy NER:**

**1. Label inventory:**
NLTK uses coarse labels (ORGANIZATION, PERSON, GPE, FACILITY).
spaCy uses a richer set: ORG, PERSON, GPE, MONEY, DATE, CARDINAL, LAW, PRODUCT, etc.
This means spaCy can label "$1.05bn" as MONEY and "August" as DATE; NLTK cannot.

**2. Multi-word entities:**
spaCy processes full sentence context and correctly groups multi-word names
(e.g., "San Jose" → one GPE, "Ice Cream Sandwich" → one PRODUCT).
NLTK works token-by-token after POS tagging, so it often splits these or misses them.

**3. Accuracy on this text:**
- spaCy correctly identifies: Apple (ORG), Samsung (ORG), US (GPE),
  San Jose (GPE), California (GPE), $1.05bn (MONEY), August (DATE), iPad (PRODUCT).
- NLTK misses monetary amounts and dates entirely, and fragments names like
  "Galaxy S III" into separate chunks.

**Verdict:** spaCy performs better. Its statistical neural model handles
domain-specific names and multi-token spans far more reliably than NLTK's
rule-based chunker.


### [points: 2] Exercise 3c: Constituency/dependency parsing
Choose one sentence from the text and run constituency parsing using NLTK and dependency parsing using spaCy.
* describe briefly the difference between constituency parsing and dependency parsing
* describe differences between the output from NLTK and spaCy.

### Answer

**Constituency vs Dependency parsing:**

*Constituency parsing* groups words into nested phrases (NP, VP, PP, S) forming a
hierarchical tree. It answers: *what are the phrase-level building blocks of the sentence?*

*Dependency parsing* draws a directed arc from each word to its syntactic head,
labelled with a grammatical relation (nsubj, dobj, prep, conj, …). It answers:
*which word governs which other word, and how?*

**Chosen sentence (index 3):**
*"In August, Samsung lost a US patent case to Apple and was ordered to pay its rival
$1.05bn (£0.66bn) in damages for copying features of the iPad and iPhone in its Galaxy
range of devices."*

**NLTK constituency output:**
The RegexpParser produces a shallow tree. It groups tokens into NP chunks
("a US patent case", "its rival") and PP chunks ("in August", "to Apple").
The coordination "lost … and was ordered" is not captured because the hand-written
grammar has no rule for coordinated VPs. Many tokens remain outside any phrase.

**spaCy dependency output:**
Every token is linked to exactly one head.
"Samsung" is the nsubj of "lost" (root). "case" is the dobj.
"Apple" is the pobj of the preposition "to". "ordered" is joined via conj.
The full sentence structure is captured, including nested PPs and the conjunction.

**Key difference:** NLTK produces a partial, flat chunking limited by the rules we write;
spaCy produces a complete word-level dependency graph for the entire sentence.


# End of this notebook